Style cleanup for S3 (#1460)
fluent/fluent-bit-docs: pipeline/outputs/s3.md

---
description: Send logs, data, and metrics to Amazon S3
---

# Amazon S3

![AWS logo](<../../.gitbook/assets/image (9).png>)

The Amazon S3 output plugin lets you ingest records into the
[S3](https://docs.aws.amazon.com/AmazonCloudWatch/latest/logs/WhatIsCloudWatchLogs.html)
cloud object store.

The plugin can upload data to S3 using the
[multipart upload API](https://docs.aws.amazon.com/AmazonS3/latest/dev/uploadobjusingmpu.html)
or [`PutObject`](https://docs.aws.amazon.com/AmazonS3/latest/API/API_PutObject.html).
Multipart is the default and is recommended. Fluent Bit will stream data in a series
of _parts_. This limits the amount of data buffered on disk at any point in time.

> By default, every time 5 MiB of data have been received, a new part will be uploaded.

Reviewer: ```suggestion
By default, every time 5&nbsp;MiB of data have been received, a new part will be uploaded.
```

The plugin can create files up to gigabytes in size from many small chunks or parts
using the multipart API. All aspects of the upload process are configurable.

The plugin lets you specify a maximum file size, and a timeout for uploads. A
file will be created in S3 when the maximum size or the timeout is reached, whichever
comes first.

Records are stored in files in S3 as newline delimited JSON.

See [AWS
Credentials](https://github.com/fluent/fluent-bit-docs/tree/43c4fe134611da471e706b0edb2f9acd7cdfdbc3/administration/aws-credentials.md)
for details about fetching AWS credentials.

{% hint style="info" %}
The [Prometheus success/retry/error metrics values](administration/monitoring.md)

> output by Fluent Bit's built-in http server are meaningless for S3 output. S3 has

Reviewer: ```suggestion
output by the built-in http server in Fluent Bit are meaningless for S3 output. S3 has
```

its own buffering and retry mechanisms. The Fluent Bit AWS S3 maintainers apologize
for this feature gap; you can [track our progress fixing it on GitHub](https://github.com/fluent/fluent-bit/issues/6141).
{% endhint %}

## Configuration Parameters

| Key                | Description                       | Default     |
|--------------------| --------------------------------- | ----------- |
| `region`           | The AWS region of your S3 bucket. | `us-east-1` |
| `bucket`           | S3 Bucket name                    | _none_ |
| `json_date_key`    | Specify the time key name in the output record. To disable the time key, set the value to `false`. | `date` |

> | `json_date_format` | Specify the format of the date. Supported formats are `double`, `epoch`, `iso8601` ( 2018-05-30T09:39:52.000681Z) and `_java_sql_timestamp_` (2018-05-30 09:39:52.000681) | `iso8601` |

Reviewer: ```suggestion
| `json_date_format` | Specify the format of the date. Accepted values: `double`, `epoch`, `iso8601` (2018-05-30T09:39:52.000681Z), `_java_sql_timestamp_` (2018-05-30 09:39:52.000681). | `iso8601` |
```

| `s3_key_format_tag_delimiters` | A series of characters used to split the tag into parts for use with `s3_key_format`. option. | `.` |
| `static_file_path` | Disables behavior where UUID string appendeds to the end of the S3 key name when `$UUID` is not provided in `s3_key_format`. `$UUID`, time formatters, `$TAG`, and other dynamic key formatters all work as expected while this feature is set to true. | `false` |
| `use_put_object`   | Use the S3 `PutObject` API instead of the multipart upload API. When enabled, the key extension is only available when `$UUID` is specified in `s3_key_format`. If `$UUID` isn't included, a random string appends format string and the key extension can't be customized. | `false` |
| `role_arn` | ARN of an IAM role to assume (for example, for cross account access.) | _none_ |
| `endpoint`  | Custom endpoint for the S3 API. Endpoints can contain scheme and port. | _none_ |

> | `sts_endpoint` | Custom endpoint for the STS API.        | _none_            |

Reviewer: ```suggestion
| `sts_endpoint` | Custom endpoint for the STS API.        | _none_ |
```

| `auto_retry_requests` | Immediately retry failed requests to AWS services once. This option doesn't affect the normal Fluent Bit retry mechanism with backoff. Instead, it enables an immediate retry with no delay for networking errors, which may help improve throughput during transient network issues. | `true` |
| `log_key` | By default, the whole log record will be sent to S3. When specifing a key name with this option, only the value of that key sends to S3. For example, when using Docker you can specify `log_key log` and only the log message sends to S3. | _none_ |
| `preserve_data_ordering` | When an upload request fails, the last received chunk might swap with a later chunk, resulting in data shuffling. This feature prevents shuffling by using a queue logic for uploads. | `true` |
| `storage_class` | Specify the [storage class](https://docs.aws.amazon.com/AmazonS3/latest/API/API_PutObject.html#AmazonS3-PutObject-request-header-StorageClass) for S3 objects. If this option isn't specified, objects store with the default `STANDARD` storage class. | _none_ |

> | `retry_limit` | Integer value to set the maximum number of retries allowed. Requires versions 1.9.10 and 2.0.1 or higher. For previous version, the number of retries is 5 and isn;t configurable. | `1` |

Reviewer: ```suggestion
| `retry_limit` | Integer value to set the maximum number of retries allowed. Requires versions 1.9.10 and 2.0.1 or later. For previous version, the number of retries is `5` and isn't configurable. | `1` |
```

| `external_id` | Specify an external ID for the STS API. Can be used with the `role_arn` parameter if your role requires an external ID. | _none_ |
| `workers` | The number of [workers](../../administration/multithreading.md#outputs) to perform flush operations for this output. | `1` |

## TLS / SSL

To skip TLS verification, set `tls.verify` as `false`. For more details about the

>  properties available and general configuration, refer to
>  [TLS/SSL](../../administration/transport-security.md).

Reviewer: ```suggestion
properties available and general configuration, refer to
[TLS/SSL](../../administration/transport-security.md).
```

## Permissions

The plugin requires the following AWS IAM permissions:

```text
{
    "Version": "2012-10-17",
    "Statement": [{
        "Effect": "Allow",
        "Action": [
            "s3:PutObject"
        ],
        "Resource": "*"
    }]
}
```

## Differences between S3 and other Fluent Bit outputs

The S3 output plugin is used to upload large files to an Amazon S3 bucket, while
most other outputs which send many requests to upload data in batches of a few
megabytes or less.

> When Fluent Bit recieves logs it stores them in chunks, either in memory or the

Reviewer: ```suggestion
When Fluent Bit receives logs, it stores them in chunks, either in memory or the
```

> filesystem depending on your settings. Chunks are usually around 2 MB in size.

Reviewer: ```suggestion
filesystem depending on your settings. Chunks are usually around 2&nbsp;MB in size.
```

Fluent Bit sends chunks, in order, to each output that matches their tag. Most outputs
then send the chunk immediately to their destination. A chunk is sent to the output's
`flush` callback function, which must return one of `FLB_OK`, `FLB_RETRY`, or
`FLB_ERROR`. Fluent Bit keeps count of the return values from each output's
`flush` callback function. These counters are the data source for Fluent Bit's error, retry,
and success metrics available in Prometheus format through its monitoring interface.

The S3 output plugin conforms to the Fluent Bit output plugin specification.

> Since S3's use case is to upload large files (over 2MB), its behavior is different.

Reviewer: ```suggestion
Since S3's use case is to upload large files (over 2&nbsp;MB), its behavior is different.
```

parameter is not meaningful for S3 since it has its own buffering system in the
`store_dir`. Instead, use `store_dir_limit_size`. S3 requires a writeable filesystem.
Running Fluent Bit on a read-only filesystem won't work with the S3 output.

S3 uploads primarily initiate using the S3
[`timer`](https://docs.aws.amazon.com/iotevents/latest/apireference/API_iotevents-data_Timer.html)
callback function, which runs separately from its `flush`.

S3 has its own buffering system and its own callback to upload data, so the normal
sequential data ordering of chunks provided by the Fluent Bit engine may be
compromised. S3 has the `presevere_data_ordering` option which ensures data is
uploaded in the original order it was collected by Fluent Bit.

### Summary: Uniqueness in S3 Plugin

- The HTTP Monitoring interface output metrics are not meaningful for S3. AWS
  understands that this is non-ideal; we have
  [opened an issue with a design](https://github.com/fluent/fluent-bit/issues/6141)
  to allow S3 to manage its own output metrics.
- You must use `store_dir_limit_size` to limit the space on disk used by S3 buffer files.

> - The original ordering of data inputted to Fluent Bit may not be preserved unless you enable `preserve_data_ordering On`.

Reviewer: ```suggestion
- The original ordering of data inputted to Fluent Bit may not be preserved unless you enable
`preserve_data_ordering On`.
```

## S3 Key Format and Tag Delimiters

In Fluent Bit, all logs have an associated tag. The `s3_key_format` option lets you
inject the tag into the S3 key using the following syntax:

- `$TAG`: The full tag.
- `$TAG[n]`: The nth part of the tag (index starting at zero). This syntax is copied
  from the rewrite tag filter. By default, “parts” of the tag are separated with
  dots, but you can change this with `s3_key_format_tag_delimiters`.

In the following example, assume the date is January 1st, 2020 00:00:00 and the tag
associated with the logs in question is `my_app_name-logs.prod`.

```python
[OUTPUT]
    Name  s3
    Match *
    bucket                       my-bucket
    region                       us-west-2
    total_file_size              250M
    s3_key_format                /$TAG[2]/$TAG[0]/%Y/%m/%d/%H/%M/%S/$UUID.gz
    s3_key_format_tag_delimiters .-
```

> With the delimiters as `.` and `-,` the tag splits into parts as follows:

Reviewer: ```suggestion
With the delimiters as `.` and `-`, the tag splits into parts as follows:
```

- `$TAG[0]` = `my_app_name`
- `$TAG[1]` = `logs`
- `$TAG[2]` = `prod`

The key in S3 will be `/prod/my_app_name/2020/01/01/00/00/00/bgdHN1NM.gz`.

### Allowing a file extension in the S3 Key Format with $UUID

The Fluent Bit S3 output was designed to ensure that previous uploads will never be

> over-written by a subsequent upload. The `s3_key_format` supports time formatters,

Reviewer: ```suggestion
overwritten by a subsequent upload. The `s3_key_format` supports time formatters,
```

> `$UUID`, and `$INDEX`. `$INDEX` is special because it is saved in the `store_dir`; if
> you restart Fluent Bit with the same disk, then it can continue incrementing the

Reviewer: ```suggestion
`$UUID`, and `$INDEX`. `$INDEX` is special because it is saved in the `store_dir`. If
you restart Fluent Bit with the same disk, it can continue incrementing the
```

index from its last value in the previous run.

For files uploaded with the `PutObject` API, the S3 output requires that a unique
random string be present in the S3 key. Many of the use cases for
`PutObject` uploads involve a short time period between uploads, so a timestamp
in the S3 key may not be unique enough between uploads. For example, if you only
specify minute granularity timestamps in the S3 key, with a small upload size, it is
possible to have two uploads that have timestamps set in the same minute. This
requirement can be disabled with `static_file_path On`.

> There are three cases where the `PutObject` API is used:

Reviewer: ```suggestion
The `PutObject` API is used in these cases:
```

> 1. When you explicitly set `use_put_object On`.
> 1. On startup when the S3 output finds old buffer files in the `store_dir` from
>    a previous run and attempts to send all of them at once.
> 1. On shutdown. To prevent data loss the S3 output attempts to send all currently
>    buffered data at once.

Reviewer: ```suggestion
- When you explicitly set `use_put_object On`.
- On startup when the S3 output finds old buffer files in the `store_dir` from
  a previous run and attempts to send all of them at once.
- On shutdown. To prevent data loss the S3 output attempts to send all currently
  buffered data at once.
```

You should always specify `$UUID` somewhere in your S3 key format. Otherwise, if the

> `PutObject` API is used, S3 appends a random 8 character UUID to the end of your

Reviewer: ```suggestion
`PutObject` API is used, S3 appends a random eight-character UUID to the end of your
```

S3 key. This means that a file extension set at the end of an S3 key will have the
random UUID appended to it. Disabled this with `static_file_path On`.

> For example, we attempt to set a `.gz` extension without specifying `$UUID`.

Reviewer: ```suggestion
For example, we attempt to set a `.gz` extension without specifying `$UUID`:
```

    s3_key_format                /$TAG/%Y/%m/%d/%H_%M_%S.gz
```

In the case where pending data is uploaded on shutdown, if the tag was `app`, the S3
key in the S3 bucket might be:

```text
/app/2022/12/25/00_00_00.gz-apwgylqg
```

The S3 output appended a random string to the file extension, since this upload
on shutdown used the `PutObject` API.

> There are two ways of disabling this behavior.

Reviewer: ```suggestion
There are two ways of disabling this behavior:
```

    ```python
    [OUTPUT]
        Name  s3
        Match *
        bucket                       my-bucket
        region                       us-west-2
        total_file_size              50M
        use_put_object               Off
        compression                  gzip
        s3_key_format                /$TAG/%Y/%m/%d/%H_%M_%S.gz
        static_file_path             On
    ```

- Explicitly define where the random UUID will go in the S3 key format:

    ```python
    [OUTPUT]
        Name  s3
        Match *
        bucket                       my-bucket
        region                       us-west-2
        total_file_size              50M
        use_put_object               Off
        compression                  gzip
        s3_key_format                /$TAG/%Y/%m/%d/%H_%M_%S/$UUID.gz
    ```

## Reliability

The `store_dir` is used to temporarily store data before upload. If Fluent Bit
stops suddenly, it will try to send all data and complete all uploads before it
shuts down. If it can not send some data, on restart it will look in the `store_dir`
for existing data and try to send it.

Multipart uploads are ideal for most use cases because they allow the plugin to

> upload data in small chunks over time. For example, 1 GB file can be created from 200
> 5MB chunks. While the file size in S3 will be 1 GB, only 5 MB will be buffered on
> disk at any one point in time.

Reviewer: ```suggestion
upload data in small chunks over time. For example, 1&nbsp;GB file can be created
from 200 5&nbsp;MB chunks. While the file size in S3 will be 1&nbsp;GB, only
5&nbsp;MB will be buffered on disk at any one point in time.
```

  in the `store_dir` longer than the configured `upload_timeout`, it will be sent
  even when the desired byte size hasn't been reached.
  If you configure a small `upload_timeout`, your files may be smaller
  than the `total_file_size`. The timeout is evaluated against the time at which S3
  started buffering data for each unqiue tag (that is, the time when new data was
  buffered for the unique tag after the last upload). The timeout is also evaluated
  against the
  [CreateMultipartUpload](https://docs.aws.amazon.com/AmazonS3/latest/API/API_CreateMultipartUpload.html)
  time, so a multipart upload will be completed after `upload_timeout` has elapsed,
  even if the desired size has not yet been reached.

If your `upload_timeout` triggers an upload before the pending buffered data reaches
the `upload_chunk_size`, it may be too small for a multipart upload. S3 will
fallback to use the [`PutObject` API](https://docs.aws.amazon.com/AmazonS3/latest/API/API_PutObject.html).

When you enable compression, S3 applies the compression algorithm at send time. The
size settings trigger uploads based on the size of buffered data, not the

> final compressed size. It is possible that after compression, buffered data no longer

Reviewer: ```suggestion
final compressed size. It's possible that after compression, buffered data no longer
```

meets the required minimum S3
[UploadPart](https://docs.aws.amazon.com/AmazonS3/latest/API/API_UploadPart.html)
size. If this occurs, you will see a log message like:

```text
[ info] [output:s3:s3.0] Pre-compression upload_chunk_size= 5630650, After
compression, chunk is only 1063320 bytes, the chunk was too small, using PutObject to upload
```

If you encounter this frequently, use the numbers in the messages to guess your
compression factor. In this example, the buffered data was reduced from

> 5,630,650 bytes to 1,063,320 bytes. The compressed size is 1/5 the actual data size.

Reviewer: ```suggestion
5,630,650 bytes to 1,063,320 bytes. The compressed size is one-fifth the actual data size.
```

AWS credentials:

```text
aws ecr-public get-login-password --region us-east-1 | docker login --username AWS --password-stdin public.ecr.aws
```

See the
[Amazon ECR Public official documentation](https://docs.aws.amazon.com/AmazonECR/latest/public/get-set-up-for-amazon-ecr.html)
for more details.

### Docker Hub

[amazon/aws-for-fluent-bit](https://hub.docker.com/r/amazon/aws-for-fluent-bit/tags)
is also available from the Docker Hub.

### Amazon ECR

Use our SSM Public Parameters to find the Amazon ECR image URI in your region:

```text
aws ssm get-parameters-by-path --path /aws/service/aws-for-fluent-bit/
```

For more information, see the
[AWS for Fluent Bit GitHub repo](https://github.com/aws/aws-for-fluent-bit#public-images).

## Advanced usage

### Use Apache Arrow for in-memory data processing

With Fluent Bit v1.8 or greater, the Amazon S3 plugin includes the support for
[Apache Arrow](https://arrow.apache.org/). Support isn't enabled by
default, and has a dependency on a shared version of `libarrow`.

To use this feature, `FLB_ARROW` must be turned on at compile time. Use the following
commands:

```text
cd build/
cmake -DFLB_ARROW=On ..
cmake --build .
```

> Once compiled, Fluent Bit can upload incoming data to S3 in Apache Arrow format.

Reviewer: ```suggestion
After being compiled, Fluent Bit can upload incoming data to S3 in Apache Arrow format.
```
